### PR TITLE
Add mask_restore to restore images based on mask, fixing #665

### PR DESCRIPTION
**Pull request by xaedes**
Add mask_restore option to give users the option to restore images based on mask, fixing #665.

Before commit c73fdd78  (Implement masking during sampling to improve blending, #308) image mask was applied after sampling, resulting in masked parts that are not regenerated to actually stay the same.
Since c73fdd78 the masked img2img will change the whole image, even in masked areas. It gives better looking results at first glance, but will result in image degredation when applied a few times. See issue #665.

In the workflow of using repeated masked img2img, users may want to use this options to keep the parts of image they actually want to keep without image degradation. A final masked img2img or whole image img2img with mask_restore disabled will give the better blending of "Implement masking during sampling".

### CODE REVIEW

File: frontend/frontend.py [hunk 1]
```diff
@@ -197,9 +197,13 @@ def draw_gradio_ui(opt, img2img=lambda x: x, txt2img=lambda x: x, imgproc=lambda
                                                             value=img2img_mask_modes[img2img_defaults['mask_mode']],
                                                             visible=True)
 
-                                    img2img_mask_blur_strength = gr.Slider(minimum=1, maximum=10, step=1,
+                                    img2img_mask_restore = gr.Checkbox(label="Restore image by mask",
```
Comment (codedealer) [inline frontend/frontend.py:200]:
```suggestion
                                    img2img_mask_restore = gr.Checkbox(label="Restore masked image", 
```
This probably needs a more user friendly name.

**Review by codedealer — CHANGES_REQUESTED** [review]
Please fix **ui_functions.change_image_editor_mode** by adding `gr.update` to the end of the list it returns (keep in mind there are two return statements there).

File: scripts/webui.py [hunk 2]
```diff
@@ -2099,10 +2124,39 @@ def run_RealESRGAN(image, model_name: str):
...
+        update_image_editor,
+        update_image_mask,
+        update_btn_editor,
+        update_btn_mask,
+        update_painterro_btn,
+        update_mask,
+        update_mask_blur_strength,
+        update_mask_restore,
+        # unknown,
+    ]
```
[the remark this answers is not in the document]
Comment (xaedes) [reply scripts/webui.py:2159]: I appended the `gr.update(..)` calls to the end of the returned lists. From the code it was totally unclear which entry in the list was for which component, so I gave each gr.update the name indicating the corresponding updated component and then return the list of named updates.

But I do understand that this is maybe a bit too much restructure for this PR. I will revert and just add gr.update at the end of the lists like you said in the earlier comment.

File: frontend/frontend.py (hunk 1 above)
Comment (xaedes) [reply frontend/frontend.py]: Good point. I changed it to "Only modify regenerated parts of image".

File: scripts/webui.py [hunk 3]
```diff
@@ -2099,10 +2124,10 @@ def run_RealESRGAN(image, model_name: str):
 img2img_toggle_defaults = [img2img_toggles[i] for i in img2img_defaults['toggles']]
 img2img_image_mode = 'sketch'
 
-def change_image_editor_mode(choice, cropped_image, resize_mode, width, height):
+def change_image_editor_mode(choice, cropped_image, mask, resize_mode, width, height):
```
Comment (xaedes) [inline scripts/webui.py:2127]: Note: here was an argument missing, I inserted it. It was no issue though, as the arguments are not used anyway.

File: frontend/frontend.py [hunk 4]
```diff
@@ -290,8 +294,14 @@ def draw_gradio_ui(opt, img2img=lambda x: x, txt2img=lambda x: x, imgproc=lambda
...
-                    [img2img_image_editor, img2img_image_mask, img2img_btn_editor, img2img_btn_mask,
-                     img2img_painterro_btn, img2img_mask, img2img_mask_blur_strength]
+                    [img2img_image_editor, 
+                     img2img_image_mask,
+                     img2img_btn_editor, 
+                     img2img_btn_mask,
+                     img2img_painterro_btn, 
+                     img2img_mask, 
+                     img2img_mask_blur_strength, 
+                     img2img_mask_restore]
```
[the remark this answers is not in the document]
Comment (xaedes) [reply frontend/frontend.py]: There I appended the new `img2img_mask_restore` component to the outputs list of the `change_image_editor_mode` function, where the gr.update is also added the end of the list to also update the new `img2img_mask_restore` component.

**Comment by xaedes** [comment]
Ohhh I see, there are two `change_image_editor_mode` functions! 
The one that I changed in webui.py seems to be an older version of the correct one that seems to be nowhere used.  
I now changed the one in ui_functions.py

**Comment by codedealer** [comment]
@xaedes remove the unused one from webui please as well. I have no idea where exactly it was introduced (I thought it was you initially) but it proves that it causes much confusion.

**Comment by xaedes** [comment]
@codedealer 
The two the functions that directly follow in code: `update_image_mask` and `copy_img_to_upscale_esrgan` are also unused duplicates of functions in `ui_functions.py`. 
I guess it makes sense to remove them as well. I will commit with all three duplicate functions removed from webui.py

**Review by codedealer — APPROVED** [review]
I'm not sure it should be on by default. It gives poor results on a simple img2img run without loopback. But we leave it on for now 👍

**Pull request merged**